File: day25/src/main/python/solution.py
if __name__ == "__main__":
    grid = {}
    y = 0
    x_max = 0
    y_max = 0
    f = open('../resources/input', 'r')
    for line in f.readlines():
        for x in range(0, len(line.strip())):
            if line[x] != '.':
                grid[(x, y)] = line[x]
        x_max = len(line) - 1
        y += 1
    y_max = y
    changed = True
    count = 1
    while changed:
        count += 1
        changed = False
        newGrid = grid.copy()
        for x, y in grid:
            value = grid[(x, y)]
            if value == '>' and ((x + 1) % x_max, y) not in grid.keys():
                newGrid.pop((x, y))
                newGrid[((x + 1) % x_max, y)] = '>'
                changed = True
        grid = newGrid.copy()
        for x, y in grid:
            value = grid[(x, y)]
            if value == 'v' and (x, (y + 1) % y_max) not in grid.keys():
                newGrid.pop((x, y))
                newGrid[(x, (y + 1) % y_max)] = 'v'
                changed = True
        grid = newGrid
    print("part 1")
    print(count)

File: day25/src/main/cpp/solution.cpp
#include <fstream>
#include <iostream>
#include <map>

using namespace std;
#define coordinate pair<int, int>
int yMax = 0;
int xMax = 0;

map<coordinate, char> readInputFile() {
  ifstream file("../resources/input");
  string s;
  auto y = 0;
  map<coordinate, char> grid;
  while (getline(file, s)) {
    for (int x = 0; x < s.length(); ++x) {
      if (s[x] != '.') {
        grid[coordinate(x, y)] = s[x];
      }
    }
    ++y;
    xMax = s.length();
  }
  yMax = y;
  return grid;
}

int main() {
  auto grid = readInputFile();
  bool changed = true;
  int count = 0;
  while (changed) {
    changed = false;
    auto next = grid;
    auto it = grid.begin();
    while (it != grid.end()) {
      if (it->second == '>' &&
          grid.find(coordinate((it->first.first + 1) % xMax,
                               it->first.second)) == grid.end()) {
        next[coordinate((it->first.first + 1) % xMax, it->first.second)] = '>';
        next.erase(it->first);
        changed = true;
      }
      it++;
    }
    grid = next;
    it = grid.begin();
    while (it != grid.end()) {
      if (it->second == 'v' &&
          grid.find(coordinate(it->first.first,
                               (it->first.second + 1) % yMax)) == grid.end()) {
        next[coordinate(it->first.first, (it->first.second + 1) % yMax)] = 'v';
        next.erase(it->first);
        changed = true;
      }
      it++;
    }
    grid = next;
    ++count;
    if (count > 1000) {
      changed = false;
    }
  }
  cout << "part 1" << endl << count << endl;
  return 0;
}